### PR TITLE
Fix `components-e2e` Pipeline

### PR DESCRIPTION
**Pull request by TanayParikh**
Support for the `QuarantinedTest` attribute.

Using the `Quarantined` trait defined here:
https://github.com/dotnet/aspnetcore/blob/ba18614ec220c1209a3976f66aacc2c761568931/src/Testing/src/xunit/QuarantinedTestTraitDiscoverer.cs#L18

Extension of https://github.com/dotnet/aspnetcore/pull/38693 which got auto-merged.

### CODE REVIEW

**Comment by pranavkm** [comment]
Out of curiosity, where do quarantined Blazor e2e tests run?

**Comment by TanayParikh** [comment]
> Out of curiosity, where do quarantined Blazor e2e tests run?

They're now being run as part of a new step of the `components-e2e` pipeline:

https://github.com/dotnet/aspnetcore/pull/38693/files#diff-eb7fafe7b73cc5c0d1cf18055a0392f032ef78a1e9a46ec7996448f07a4aba16R49-R51

**Review by pranavkm — APPROVED** [review]

File: .azure/pipelines/components-e2e-tests.yml [hunk 1]
```diff
@@ -44,9 +44,9 @@ jobs:
       displayName: NPM install
     - script: .dotnet/dotnet build ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-restore
       displayName: Build
-    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter QuarantinedTest\!~github --logger trx
+    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter Quarantined\!~github --logger trx
       displayName: Run E2E tests
-    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter QuarantinedTest~github -p:RunQuarantinedTests=true --logger trx
+    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter Quarantined~github -p:RunQuarantinedTests=true --logger trx
```
Comment (dougbu) [inline .azure/pipelines/components-e2e-tests.yml]: Compare to
- https://github.com/dotnet/aspnetcore/blob/main/eng/helix/content/RunTests/TestRunner.cs#L225
- https://github.com/dotnet/aspnetcore/blob/main/eng/helix/content/RunTests/TestRunner.cs#L243

It's likely you're still using the wrong syntax and `$(RunQuarantinedTests)` overrides `--filter` options anyhow: https://github.com/dotnet/aspnetcore/blob/main/eng/QuarantinedTests.BeforeArcade.props#L7-L10

File: .azure/pipelines/components-e2e-tests.yml [hunk 2]
```diff
@@ -44,9 +44,9 @@ jobs:
       displayName: NPM install
     - script: .dotnet/dotnet build ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-restore
       displayName: Build
-    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter QuarantinedTest\!~github --logger trx
+    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter Quarantined\!~github --logger trx
```
Comment (dougbu) [inline .azure/pipelines/components-e2e-tests.yml]:
```suggestion
    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter 'Quarantined!=true|Quarantined=false' --logger trx
```

File: .azure/pipelines/components-e2e-tests.yml (hunk 1 above)
Comment (dougbu) [inline .azure/pipelines/components-e2e-tests.yml]:
```suggestion
    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter 'Quarantined=true' -p:RunQuarantinedTests=true --logger trx
```

File: .azure/pipelines/components-e2e-tests.yml [hunk 3]
```diff
@@ -44,9 +44,9 @@ jobs:
       displayName: NPM install
     - script: .dotnet/dotnet build ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-restore
       displayName: Build
-    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter QuarantinedTest\!~github --logger trx
+    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter Quarantined\!=true --logger trx
```
Comment (dougbu) [inline .azure/pipelines/components-e2e-tests.yml:47]: Missed an update, sorry
```suggestion
    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter 'Quarantined!=true|Quarantined=false' --logger trx
```

Comment (dougbu) [reply .azure/pipelines/components-e2e-tests.yml:47]: Note: I'm not positive this is necessary. It's also unclear why we are using different options in eng\QuarantinedTests.BeforeArcade.props and eng\helix\content\RunTests\TestRunner.cs. @HaoK❔

Comment (TanayParikh) [reply .azure/pipelines/components-e2e-tests.yml:47]: Given this is linux/macos, I believe the `\` is required for `!`.

Comment (TanayParikh) [reply .azure/pipelines/components-e2e-tests.yml:47]: Pipeline is now passing, going to merge as is without this additional change, let me know if you disagree.

https://dev.azure.com/dnceng/public/_build/results?buildId=1489638&view=logs&j=fe94c0c9-bb8c-5d6f-3b51-887173cc2f5c&t=b4bd3469-f776-5941-b623-568498d38bde

Comment (dougbu) [reply .azure/pipelines/components-e2e-tests.yml:47]: Results look _almost_ right. Unfortunately, the `Quarantined=true` run put it's results in the same directory as this one. Probably need to add something more to that command line to make it possible for the second test results upload to work and to make the Tests tab easier to read.

Comment (dougbu) [reply .azure/pipelines/components-e2e-tests.yml:47]: (Root cause of problems here is use of `dotnet test` on build agents. We use Arcade infrastructure, including `$(ArtifactsTestResultsDir)` support everywhere else.)

File: .azure/pipelines/components-e2e-tests.yml [hunk 4]
```diff
@@ -44,9 +44,9 @@ jobs:
...
       displayName: Run E2E tests
-    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter QuarantinedTest~github -p:RunQuarantinedTests=true --logger trx
+    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter Quarantined=true -p:RunQuarantinedTests=true --logger trx
```
Comment (dougbu) [inline .azure/pipelines/components-e2e-tests.yml:49]:
```suggestion
    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter Quarantined=true -p:RunQuarantinedTests=true --logger trx --logger:"trx%3BLogFileName=$(_TestResultTrxFileName)" --logger:"html%3BLogFileName=$(_TestResultHtmlFileName)" "--ResultsDirectory:$(_TestResultDirectory)"`
```

Up to you whether the HTML logger is needed and how you want to set the filenames. See src/Microsoft.DotNet.Arcade.Sdk/tools/VSTest.targets and src/Microsoft.DotNet.Arcade.Sdk/tools/Tests.targets in the Arcade repo for inspiration.

**Review by dougbu — CHANGES_REQUESTED** [review]

File: .azure/pipelines/components-e2e-tests.yml (hunk 3 above)
Comment (dougbu) [reply .azure/pipelines/components-e2e-tests.yml:47]: > Given this is linux, I believe the `\` is required for `!`.

nit: You ignored the single quotes in my suggestion.

File: .azure/pipelines/components-e2e-tests.yml [hunk 5]
```diff
@@ -44,9 +44,9 @@ jobs:
       displayName: NPM install
     - script: .dotnet/dotnet build ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-restore
       displayName: Build
-    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter QuarantinedTest\!~github --logger trx
+    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter 'Quarantined!=true|Quarantined=false' --logger trx
       displayName: Run E2E tests
-    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter QuarantinedTest~github -p:RunQuarantinedTests=true --logger trx
+    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter 'Quarantined=true' -p:RunQuarantinedTests=true --logger trx --logger:"trx%3BLogFileName=$(_TestResultTrxFileName)" --logger:"html%3BLogFileName=$(_TestResultHtmlFileName)" "--ResultsDirectory:$(_TestResultDirectory)"`
```
Comment (dougbu) [inline .azure/pipelines/components-e2e-tests.yml:49]: I doubt this command will work without setting `$(_TestResultTrxFileName)`, `$(_TestResultHtmlFileName)`, and `$(_TestResultDirectory)`. My suggestion wasn't meant to be complete (see comments with it).

File: .azure/pipelines/components-e2e-tests.yml [hunk 6]
```diff
@@ -44,9 +44,9 @@ jobs:
       displayName: NPM install
     - script: .dotnet/dotnet build ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-restore
       displayName: Build
-    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter QuarantinedTest\!~github --logger trx
+    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter 'Quarantined!=true|Quarantined=false' --logger trx
```
Comment (dougbu) [inline .azure/pipelines/components-e2e-tests.yml:47]: While here, let's clean up the results file names and add HTML results. Suggestion indentation looks a bit odd however.

```suggestion
    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter 'Quarantined!=true|Quarantined=false'
                 --logger:"trx%3BLogFileName=Microsoft.AspNetCore.Components.E2ETests.trx"
                 --logger:"html%3BLogFileName=Microsoft.AspNetCore.Components.E2ETests.html"
                 "--ResultsDirectory:$(Build.SourcesDirectory)/artifacts/TestResults/$(_BuildConfig)"
```

File: .azure/pipelines/components-e2e-tests.yml (hunk 5 above)
Comment (dougbu) [reply .azure/pipelines/components-e2e-tests.yml:49]: Failed at least in part due to the backtick I accidentally included in my suggestion. How about
```suggestion
    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter 'Quarantined=true' -p:RunQuarantinedTests=true
                 --logger:"trx%3BLogFileName=Microsoft.AspNetCore.Components.E2ETests.trx"
                 --logger:"html%3BLogFileName=Microsoft.AspNetCore.Components.E2ETests.html"
                 "--ResultsDirectory:$(Build.SourcesDirectory)/artifacts/TestResults/$(_BuildConfig)/Quarantined"
```

**Review by dougbu — COMMENTED** [review]
More&hellip;

**Comment by TanayParikh** [comment]
Thanks Doug, applied your suggestions (and replaced `_BuildConfig` -> `BuildConfiguration`). Will monitor CI for results.

File: .azure/pipelines/components-e2e-tests.yml [hunk 7]
```diff
@@ -44,9 +44,15 @@ jobs:
       displayName: NPM install
     - script: .dotnet/dotnet build ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-restore
       displayName: Build
-    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter QuarantinedTest\!~github --logger trx
+    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter 'Quarantined!=true|Quarantined=false'
```
Comment (dougbu) [inline .azure/pipelines/components-e2e-tests.yml:47]: I accidentally dropped `-p:RunQuarantinedTests=true` in my suggestion I think.

File: .azure/pipelines/components-e2e-tests.yml [hunk 8]
```diff
@@ -44,9 +44,15 @@ jobs:
...
+                 --logger:"trx%3BLogFileName=Microsoft.AspNetCore.Components.E2ETests.trx"
+                 --logger:"html%3BLogFileName=Microsoft.AspNetCore.Components.E2ETests.html"
+                 "--ResultsDirectory:$(Build.SourcesDirectory)/artifacts/TestResults/$(BuildConfiguration)"
```
Comment (dougbu) [inline .azure/pipelines/components-e2e-tests.yml:50]: Too much copying 'n pasting. Need to update the `searchFolder on line 63 to match this.

**Comment by dougbu** [comment]
Looks like copying from Arcade didn't help you:

``` text
MSBUILD : error MSB1001: Unknown switch.
Switch: --ResultsDirectory:/home/vsts/work/1/s/artifacts/TestResults/Release
```

**Comment by dougbu** [comment]
`dotnet test --help` says that should be `--results-directory`

File: .azure/pipelines/components-e2e-tests.yml (hunk 7 above)
Comment (TanayParikh) [reply .azure/pipelines/components-e2e-tests.yml:47]: We have this on line 52 for the quarantined runs. I don't believe we need that switch for the non-quarantine run on line 47 here.

Comment (dougbu) [reply .azure/pipelines/components-e2e-tests.yml:47]: My bad

File: .azure/pipelines/components-e2e-tests.yml [hunk 9]
```diff
@@ -44,25 +44,31 @@ jobs:
       displayName: NPM install
     - script: .dotnet/dotnet build ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-restore
       displayName: Build
-    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter QuarantinedTest\!~github --logger trx
+    - script: .dotnet/dotnet test ./src/Components/test/E2ETest -c $(BuildConfiguration) --no-build --filter 'Quarantined!=true|Quarantined=false'
+                 --logger:"trx%3BLogFileName=Microsoft.AspNetCore.Components.E2ETests.trx"
+                 --logger:"html%3BLogFileName=Microsoft.AspNetCore.Components.E2ETests.html"
+                 --results-directory $(Build.SourcesDirectory)/artifacts/TestResults/$(BuildConfiguration)/Unquarantined
```
Comment (dougbu) [inline .azure/pipelines/components-e2e-tests.yml:50]: nit: We don't include /Unquarantined elsewhere but I guess it doesn't matter much if the artifacts are a bit different from those we see in other non-quarantined pipelines.

**Review by dougbu — APPROVED** [review]

**Comment by TanayParikh** [comment]
Screen Shot 2021-11-30 at 12 19 03 PM

🎉

**Comment by TanayParikh** [comment]
@dotnet/aspnet-build can this please be merged, it's failing on unrelated tests (quarantined here https://github.com/dotnet/aspnetcore/pull/38733 / https://github.com/dotnet/aspnetcore/pull/38737). This CI failure is in turn blocking other PRs.

**Pull request merged**